backend/cfg/cfgize.ml

```diff
@@ -569,110 +569,120 @@ let update_trap_handler_blocks : State.t -> Cfg.t -> unit =
-        stack, Some (handler_label, handler_stack))
-    else stack, exceptional_successor
+  let compute_stack_offset ~stack_offset ~traps =
+    stack_offset + (Proc.trap_size_in_bytes * List.length traps)
+
+  let check_and_set_stack_offset :
+      'a Cfg.instruction ->
+      stack_offset:int ->
+      traps:handler_stack ->
+      'a Cfg.instruction =
+   fun instr ~stack_offset ~traps ->
+    assert (instr.stack_offset = invalid_stack_offset);
+    Cfg.set_stack_offset instr (compute_stack_offset ~stack_offset ~traps)
 
   let process_terminator :
-      handler_option ->
-      handler_stack ->
+      stack_offset:int ->
+      traps:handler_stack ->
       Cfg.terminator Cfg.instruction ->
-      (handler_stack * handler_option) * Cfg.terminator Cfg.instruction =
-   fun exceptional_successor stack term ->
-    let term = Cfg.set_trap_depth term (succ (List.length stack)) in
+      int * handler_stack * Cfg.terminator Cfg.instruction =
+   fun ~stack_offset ~traps term ->
+    assert (term.stack_offset = invalid_stack_offset);
+    let term = check_and_set_stack_offset term ~stack_offset ~traps in
     match term.desc with
+    | Tailcall (Self _) when List.length traps <> 0 || stack_offset <> 0 ->
+      Misc.fatal_error
+        "Cfgize.Trap_depth_and_exn.terminator: unexpected handler on self \
+         tailcall"
+    | Tailcall (Self _)
     | Never | Return
     | Tailcall (Func _)
     | Call_no_return _ | Raise _ | Always _ | Parity_test _ | Truth_test _
     | Float_test _ | Int_test _ | Switch _ ->
-      ( record_handler stack exceptional_successor
-          ~can_raise:(Cfg.can_raise_terminator term.desc),
-        term )
-    | Tailcall (Self _) ->
-      if List.length stack <> 0
-      then
-        Misc.fatal_error
-          "Cfgize.Trap_depth_and_exn.terminator: unexpected handler on self \
-           tailcall"
-      else (stack, exceptional_successor), term
-
-  let process_basic :
-      handler_option ->
-      handler_stack ->
-      Cfg.basic Cfg.instruction ->
-      (handler_stack * handler_option) * Cfg.basic Cfg.instruction =
-   fun exceptional_successor stack instr ->
-    let instr = Cfg.set_trap_depth instr (succ (List.length stack)) in
-    match instr.desc with
-    | Pushtrap { lbl_handler } ->
-      (lbl_handler :: stack, exceptional_successor), instr
-    | Poptrap -> begin
-      match stack with
-      | [] ->
-        Misc.fatal_error
-          "Cfgize.Trap_depth_and_exn.basic: trying to pop from an empty stack"
-      | _ :: stack -> (stack, exceptional_successor), instr
-    end
-    | Op _ | Call _ | Reloadretaddr | Prologue ->
-      ( record_handler stack exceptional_successor
-          ~can_raise:(Cfg.can_raise_basic instr.desc),
-        instr )
+      stack_offset, traps, term
+
+  let rec update_block :
+      Cfg.t -> Label.t -> stack_offset:int -> traps:handler_stack -> unit =
+   fun cfg label ~stack_offset ~traps ->
+    let process_basic :
```

Comment:
(For consistency, could be moved at
the same level as `process_terminator`.)

Reply:
fixed